R/parse_variable_indices.R

```diff
@@ -0,0 +1,148 @@
+    stop_no_call(paste("Some variable names contain unpaired square brackets",
+                  "or are multi-indexed."))
+  }
+  missing_index <- grepl("\\[,|,\\]|,,", x) # check for `[,` `,,` or `,]`
+  if (any(missing_index)) {
+    stop_no_call(paste("Some variables contain missing indices. Each comma between square", 
+                       "brackets must be both preceeded and succeeded by an index."))
+  }
+
+  # Get dimensionality. Variables with no brackets are given as dimensionality zero.
+  # Variables with brackets are given dimensionality 1, even if they contain just one element.
+  dimensionality_elementwise <- sapply(vars_indices, function(x){
+    if (length(x) == 2) {
+      out <- length(strsplit(x[2], ",")[[1]]) # number of commas plus one
+    } else {
+      out <- 0
+    }
+    out
+  })
+  dimensionality <- sapply(var_names, function(x){
+    out <- unique(dimensionality_elementwise[vars == x])
+    if (length(out) != 1) {
+      stop_no_call(paste0("Inconsistent indexing found for variable ", x, " ."))
+    }
+    out
+  })
+  variable_indices_info <- lapply(var_names, function (x) {
+    # dimensions counts the number of unique values of each index.
+    # implied_dimensions gives the maximum for integer indices with no values <= 0
+    # indices is a data frame with one line per combination of indices, including implied indices
+    # position gives the position in the original vector of variables (passed to `parse_variable_indices`)
+    # where a given combination of indices is found (NA if the comination does not exist)
+    indices_info <- named_list(c("dimensionality", "dimensions", "implied_dimensions", "index_names", "indices", "position"))
+    indices_info$dimensionality <- dimensionality[[x]]
+    var_i <- vars == x
+    var_length <- sum(var_i)
+    if (dimensionality[x] == 0) {
+      # single variable, no indices
+      indices_info$dimensions <- indices_info$implied_dimensions <- NA
+      indices_info$indices <- as.integer(1)
+      dimnames(indices_info$indices) <- NULL
+      indices_info$implied_dimensions <- NA
+      indices_info$internal_position <- 1
+    } else {
+      indices <- sapply(vars_indices[var_i], `[[`, 2)
+      indices <- as.data.frame(do.call(rbind, strsplit(indices, ",")),
+                               stringsAsFactors = FALSE)
+      indices_info$dimensions <- unname(apply(indices, 2, function(x){length(unique(x))}))
+      unique_indices <- vector("list", length(indices))
+      .dimnames <- vector("list", length(indices))
+      names(unique_indices) <- names(indices)
+      for (i in seq_along(indices)) {
+        numeric_index <- suppressWarnings(as.numeric(indices[[i]]))
+        if (!anyNA(numeric_index) && rlang::is_integerish(numeric_index)) {
+          # for integer indices, we need to convert them to integers
+          # so that we can sort them in numerical order (not string order)
+          if (min(numeric_index) >= 1) {
```

Comment:
I would factor out the common code from inside both branch of this if statement and put it before it (in both branches the result should be converted to an integer):

```r
indices[[i]] <- as.integer(numeric_index)
```

Reply:
done